### PR TITLE
[runtime] Put multiple assemblies in one native library

### PR DESCRIPTION
**Pull request by alexanderkyte**
Fixes https://github.com/mono/mono/issues/6370

In certain situations it is desirable to only ship a single native
library containing all of an application's AOT code. This is done
here by creating a library which statically AOTs all of the assemblies
and embeds them in an array that the runtime can look up.

Now a design issue around finding this native library arises. We
currently discover dynamic libraries by looking for files with names
that look like the names of the relevant managed files. When you have many
managed files that are associated with a single native library,
most external solutions result in dependence on library loading order.

I thought it useful to load these static libraries using the API
endpoint that mono typically registers static native libraries with.

Usage:

Let's make two sample assemblies, one of which depends on the other
```
$> echo '''
using System;

public class FooTwo {
	public static void MainTwo (string[] args) {
		Console.WriteLine ("foobarTwo!");
	}
}
''' > FooTwo.cs
```
```
$> echo '''
public class Foo {
	public static void Main (string[] args) {
		FooTwo.MainTwo (args);
	}
}
''' > Foo.cs
```
```
$> mcs FooTwo.cs -t:library
$> mcs Foo.cs /reference:FooTwo.dll
```

Let's AOT them
```
$> mono --aot=static,full FooTwo.dll
$> mono --aot=static,full Foo.exe
```
Let's merge the two static AOT copies to make a dynamic library
```
$> mono --aot-merge=FooMerged FooTwo.dll Foo.exe
```
Let's remove the object files to make sure we're definitely getting the
code from our joined module
```
$> rm Foo.* FooTwo.*
```
Let's load the entry point in full aot mode, and watch the runtime find
AOT'ed code in the merged module
```
$> mono  --aot-preload=FooMerged.dylib --full-aot Foo.exe
```
You should see the textual output, rather than the error about
missing native code.

### CODE REVIEW

**Comment by vargaz** [comment]
Do we really need this ? The bug this fixes seems to be an extreme edge cases only encountered when running some of our tests.

**Comment by jaykrell** [comment]
I thought there was a problem where some versions of android support barely enough .sos and this moves us/clients significantly away from the limit. On windows with native code it can also be nice but not critical for debugging to have fewer dlls. Also perf can be improved via cross module linking. But mainly the first point: fewer .so on android?

**Comment by vargaz** [comment]
According to the bug report, the bug occurs on Android 4.3 which is from 2012, and it happens if the project has more than 55 assemblies.

**Comment by alexanderkyte** [comment]
The logic is rather simple compared to what the workarounds on the managed side would look like

**Comment by jonpryor** [comment]
Are there any performance implications here? If there are not, it might be interesting to consider just using it for everything. (In a FullAOT scenario, is there a benefit to having multiple separate libraries? My understanding is that iOS often requires a single binary *anyway* -- or perhaps used to? -- so a single output file might be desirable.)

If there *is* a performance implication, then it's probably not worth bothering. [Android <= 4.3](https://developer.android.com/about/dashboards/index.html) is 6.5% of devices hitting the Google Play Store (as of 2018-Feb-03), and that percentage will only fall. We would thus want to surface this as Yet Another Option in Xamarin.Android, which is already quite complex, so I'm not sure it would actually be worthwhile.

**Comment by vargaz** [comment]
Shared libraries do have some size restrictions, so it might not be possible to put everything into the same shared lib, also this makes the aot process less parallel.

**Comment by alexanderkyte** [comment]
There's no performance hit at run-time or compile-time added over what happens when mono does when we do static linking. AOT is parallel per assembly, we have a single step at the end that serially links all of these modules together. This step should be really short, it's just making a list of names and running the native linker. 

There are per-library OS limits, but I see people linking huge game libraries into applications. I don't think our applications are yet reaching those limits. If they are, it should be easy enough to make the preload option take a list of modules to load. 

Reaching some balance between library size and number of libraries should probably be left to the build system?

**Comment by alexanderkyte** [comment]
@monojenkins build

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv5

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Coop GC

**Comment by luhenry** [comment]
@monojenkins Linux ARMv5

**Comment by luhenry** [comment]
@kumpera @vargaz could you please review?

**Comment by kumpera** [comment]
We have no plans to do FullAOT on Android, so I’m at loss on why this is even a thing.

I can see the advantage of doing it across assemblies, so if this worked with regular and hybrid AOT, there would be an use for it.

Besides that, how would one use this from the embedding api?

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -10497,6 +10511,17 @@ emit_aot_file_info (MonoAotCompile *acfg, MonoAotFileInfo *info)
...
 
+static char *
+sanitize_aot_symbol_name (char *symbol)
+{
+	char *p = symbol;
+	for (p = symbol; *p; ++p) {
+		if (!(isalnum (*p) || *p == '_'))
+			*p = '_';
+	}
+	return g_strdup (symbol);
```
Comment (jaykrell) [inline mono/mini/aot-compiler.c]: Might as well strdup it first, then replace?
Also there is probably an extra albeit harmless check for '_'.

**Comment by alexanderkyte** [comment]
I'm going to close this PR and leave the branch. It can be re-opened if it's desired.
